File: utils/python/union_find.py
class UnionFind:
    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.rank = [1] * n

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])

        return self.parent[x]

    def union(self, x: int, y: int) -> None:
        x = self.find(x)
        y = self.find(y)

        if x == y:
            return

        if self.rank[x] < self.rank[y]:
            self.parent[x] = y
            self.rank[y] += self.rank[x]
        else:
            self.parent[y] = x
            self.rank[x] += self.rank[y]

File: utils/java/UnionFind.java
package utils.java;

public class UnionFind {
  private final int[] parent;
  private final int[] rank;

  public UnionFind(int n) {
    this.parent = new int[n];
    this.rank = new int[n];

    for (int i = 0; i < n; ++i) {
      parent[i] = i;
    }
  }

  public int find(int x) {
    if (parent[x] != x) {
      parent[x] = find(parent[x]);
    }

    return parent[x];
  }

  public void merge(int x, int y) {
    final int xParent = find(x);
    final int yParent = find(y);

    if (xParent == yParent) {
      return;
    }

    if (rank[xParent] < rank[yParent]) {
      parent[xParent] = yParent;
      rank[yParent] += rank[xParent];
    } else {
      parent[yParent] = xParent;
      rank[xParent] += rank[yParent];
    }
  }
}
